src/punycode.cr

```diff
@@ -0,0 +1,168 @@
+module Punycode
+  class Error < Exception; end
+
+  BASE         =  36
+  TMIN         =   1
+  TMAX         =  26
+  SKEW         =  38
+  DAMP         = 700
+  INITIAL_BIAS =  72
+  INITIAL_N    = 128
+
+  DELIMITER = '-'
+
+  BASE36 = "abcdefghijklmnopqrstuvwxyz0123456789"
+
+  private def self.adapt(delta, numpoints, firsttime)
+    delta /= firsttime ? DAMP : 2
+    delta += delta / numpoints
+    k = 0
+    while delta > ((BASE - TMIN) * TMAX) / 2
+      delta /= BASE - TMIN
+      k += BASE
+    end
+    k + (((BASE - TMIN + 1) * delta) / (delta + SKEW))
+  end
+
+  def self.encode(string : String)
+    encode string.chars
+  end
+
+  def self.encode(chars)
+    String.build { |io| encode chars, io }
+  end
+
+  def self.encode(string : String, io)
+    encode string.chars, io
+  end
+
+  def self.encode(chars, io)
+    h = 0
+    all = true
+    others = [] of Char
+
+    chars.each do |c|
+      if c < '\u0080'
+        h += 1
+        io << c
+        all = false
+      else
+        others.push c
+      end
+    end
+
+    return if others.empty?
+    others.sort!
+    io << DELIMITER unless all
+
+    delta = 0_u32
+    n = INITIAL_N
+    bias = INITIAL_BIAS
+    firsttime = true
+    prev = nil
+
+    h += 1
+    others.each do |m|
+      next if m == prev
+      prev = m
+
+      raise Error.new("overflow") if m.ord - n > (Int32::MAX - delta) / h
```

Comment:
"overflow" is a bit vague as an error message, could you be a bit more verbose here? Also ditto about custom error class.